File: Arrays/Algorithms/Sliding_Window/Two_Pointer/longest_substring_k_distinct.py
def length_of_longest_substring_k_distinct(s, k):
	from collections import defaultdict
	freq = defaultdict(int)
	l = res = 0
	for r, c in enumerate(s):
		freq[c] += 1
		while len(freq) > k:
			freq[s[l]] -= 1
			if freq[s[l]] == 0:
				del freq[s[l]]
			l += 1
		res = max(res, r - l + 1)
	return res

if __name__ == "__main__":
	s = input().strip()
	k = int(input())
	print(length_of_longest_substring_k_distinct(s, k))

File: Arrays/Algorithms/Sliding_Window/Two_Pointer/longest_substring_k_distinct.cpp
#include <bits/stdc++.h>
using namespace std;

int lengthOfLongestSubstringKDistinct(const string& s, int k) {
	unordered_map<char, int> freq;
	int l = 0, res = 0;
	for (int r = 0; r < s.size(); r++) {
		freq[s[r]]++;
		while (freq.size() > k) {
			freq[s[l]]--;
			if (freq[s[l]] == 0) freq.erase(s[l]);
			l++;
		}
		res = max(res, r - l + 1);
	}
	return res;
}

int main() {
	string s;
	int k;
	cin >> s >> k;
	cout << lengthOfLongestSubstringKDistinct(s, k) << endl;
	return 0;
}
